### ISSUE TITLE
Forced for Publication label fails in sdcTable, but works in Tau-Argus

### ISSUE THREAD

**Issue by RoelTim** [sdcTools/UserSupport#115]
SDC tool used: sdcTable & Tau-Argus
Version used: 0.27
Operating system used: Windows

When I apply the label "forced to publish (z)" the "protectTable" function is running forever with the HITAS method in R. But, the exact same problem/data does lead to a result with the Modular suppression in Tau-Argus.

Do you know what the reason could be why with the "forced for publication (z)" rule Tau-Argus does give a result while R sdcTable is not able to?

Extra detail: I apply the label "forced to publish (z)" to avoid that zeros are selected for secondary suppression. In my data certain variables are non-negative by definition and zeros using zeros for secondary suppression is not always sufficient for protection.

**Comment by bernhard-da** [comment]
hi @RoelTim 
the hitas implementation in sdcTable is different to the implementation in tau-argus. so differences are expected. i can't however look into it unless you provide a reproducable example. thx.

**Comment by ppdewolf** [comment]
Just to be sure: when you mention "zero", do you mean an "empty" cell (i.e., no contributors) or a cell with value zero (i.e., all contributors contribute zero)?
Empty cells are by default no candidates for secondary suppression in tau-argus. Cells with value zero could actually be considered "primary" unsafe (you know the contribution of all contributors).

**Comment by RoelTim** [comment]
@ppdewolf and @bernhard-da thank you a lot for fast replies. 

@bernhard-da It’s useful to know that the methods actually really differ. I will try to come back with a comparative example soon.

@ppdewolf I indeed mean cells that really have the value zero (not empty). I will increase the costs of the zeros such that will not be selected anymore for secondary suppression.

**Comment by bernhard-da** [comment]
@RoelTim please note that eventually, i will remove my implementation for HITAS from sdcTable in the future. if someone wants to use it, it is possible to create suitable inputs for tau-argus using `sdcTable::createArgusInput`

**Comment by ppdewolf** [comment]
@RoelTim please note that in tau-argus you can choose "Zero unsafe". That option will make all zero cells (value=0 AND number of contributions > 0) primary unsafe. Then you also have to specify a "safety range" to be used for those zero-cells.
Like I mentioned in my previous comment, such a cell could be considered "group disclosure" as knowing that all contributions are non-negative implies that any contributor to that cell contributes 0 (i.e., disclosure).

**Comment by RoelTim** [comment]
@berhnhard-da and @ppdewolf thank you for all your info and help.

Currently I am facing a more urgent problem. I have two very comparable datasets referring to two different time periods: one can be successfully solved in Tau-Argus while the other makes Tau-Argus crash. I cannot figure out why Tau-Argus crashes in one case and why not in the other case, especially because for the problematic period the number of confidential cells and the additivity problems are smaller. I will send you my anonymised tables to the Tau-Argus email address. Although the tables have been modified so that the original numbers cannot be retrieved I would still prefer to not share them on Github.

It would be much appreciated if you help me further.

**Comment by ppdewolf** [comment]
I'll try to look into this, but the error-log file you sent along with the problem-table shows me that there is an exception error in one of the internal functions of SCIP (the free solver). This is something I cannot easily debug. No real error-messages from that function, so also not sure if the arguments given to that function are casuing this, or if the fucntion itself is causing this.
I'll try to run the same tables using Cplex and see what happens.

**Comment by ppdewolf** [comment]
@RoelTim How do you run the non-additive tables? With the option "Compute incorrect totals" or "Allow non-additivity"?

**Comment by Incarro** [comment]
Hi @ppdewolf, I am working with @RoelTim on this project. We usually run the modular approach allowing non-additivity (the additivity problems are very small). You can find below all the settings that we use:

![image](https://user-images.githubusercontent.com/54350834/63413969-22d36e00-c3fb-11e9-8392-077ab027361a.png)

![image](https://user-images.githubusercontent.com/54350834/63413990-2ff05d00-c3fb-11e9-98a7-540cb1b6e7bc.png)

![image](https://user-images.githubusercontent.com/54350834/63414005-3a125b80-c3fb-11e9-94ad-03f899223965.png)

As @RoelTim said, one of the datasets is solved by the free solver (it takes around 190 seconds), while the other makes Tau Argus to crash. R sdcTable can solve both the problems.

**Comment by ppdewolf** [comment]
@RoelTim and @Incarro 
Running with Free solver case A gets a solution, case B crashes.
Running both tables with Xpress, I get solutions in both cases. 
Running with Cplex I get in both cases an error message stating that some subproblem is likely to be infeasible:
![afbeelding](https://user-images.githubusercontent.com/19729987/63579042-0ae22280-c592-11e9-91a7-72d5a4fcf51a.png)

But please note that when you run modular with a non-additive table you get the warning:
![afbeelding](https://user-images.githubusercontent.com/19729987/63578839-95765200-c591-11e9-89a6-21a23d1bc606.png)
So we actually warn you that there may be some problems 😉 

The problem seems to be with the non-additivity of the table. 

Moreover, note that for complex situations (lke in your case: 4 dimensions AND non-additive) we cannot guarantee that the free solver will work. In the manual we stated:

> We have also added the option to use a free open solver in addition to the classical commercial solvers like CPLEX and Xpress. However we expect that these commercial solvers are still very much needed, when we want to protect large serious tables.

**Comment by Incarro** [comment]
@ppdewolf thanks a lot for check! I tried to remove all the existing additivity problems, and the free solver seems to run without crashing (as long as the number of cells with forced for publication is not too high). However, I have noticed a strange issue with the decimals. Whenever I import an additive table in Tau Argus and I execute the additivity check, it seems that the decimals are truncated so I still have some small residual additivity problems. Any solution for this issue?

I tried the solution proposed here but it did not work:
https://github.com/sdcTools/UserSupport/issues/46

Regarding the license for Xpress, can Tau Argus support the FICO Xpress Community License?

Thanks again for the support.

**Comment by g3o2** [comment]
Not a solution but a workaround may be to transform your input into integers before tau-Argus and then back into decimals thereafter.

**Comment by Incarro** [comment]
@g3o2 indeed, at the end I solved the issue removing the decimals since are not really relevant for our problem
